### ISSUE TITLE
Issues parsing design documents with couchDB 3.1

### ISSUE THREAD

**Issue by raulmartinezr** [gesellix/couchdb-prometheus-exporter#73]
Hi,

It seems therea re some issue with design documents in couchdb version 3.1

```bash E0816 20:03:43.774989       1 collector.go:392] Error collecting stats: error collecting couchdb stats: error unmarshalling view doc for view 'uuoapzzfavifqbul/_design/portrayals/_view/index-portrayal-id': unexpected end of JSON input
E0816 20:03:48.813997       1 collector.go:392] Error collecting stats: error collecting couchdb stats: error unmarshalling view doc for view 'newcujyzimvimleh/_design/portrayals/_view/index-portrayal-id': unexpected end of JSON input
E0816 20:03:53.758340       1 collector.go:392] Error collecting stats: error collecting couchdb stats: error unmarshalling view doc for view 'wxpechanwbsesqrn/_design/accounts/_view/index-account-type-email': unexpected end of JSON input
E0816 20:03:58.768770       1 collector.go:392] Error collecting stats: error collecting couchdb stats: error unmarshalling view doc for view 'rbiorrnqhxoubndm/_design/portrayals/_view/index-portrayal-id': unexpected end of JSON input
E0816 20:04:03.772816       1 collector.go:392] Error collecting stats: error collecting couchdb stats: error unmarshalling view doc for view 'tests_given_database_when_provision_a_domain_then_resources_available/_design/portrayals/_view/index-portrayal-id': unexpected end of JSON input
E0816 20:04:08.787779       1 collector.go:392] Error collecting stats: error collecting couchdb stats: error unmarshalling view doc for view 'newcujyzimvimleh/_design/accounts/_view/index-account-type-email': unexpected end of JSON input
```

And this is how one of these documents looks: 1 document with two views

```json
{
  "_id": "_design/accounts",
  "_rev": "2-bc38cc03774e1beb53d97e494c5b6f51",
  "language": "query",
  "views": {
    "index-account-type-name": {
      "map": {
        "fields": {
          "doc_type": "asc",
          "account_name": "asc"
        },
        "partial_filter_selector": {}
      },
      "reduce": "_count",
      "options": {
        "def": {
          "fields": [
            {
              "doc_type": "asc"
            },
            {
              "account_name": "asc"
            }
          ]
        }
      }
    },
    "index-account-type-email": {
      "map": {
        "fields": {
          "doc_type": "asc",
          "email": "asc"
        },
        "partial_filter_selector": {}
      },
      "reduce": "_count",
      "options": {
        "def": {
          "fields": [
            {
              "doc_type": "asc"
            },
            {
              "email": "asc"
            }
          ]
        }
      }
    }
  },
  "options": {
    "partitioned": true
  }
}
```

Whith this error many of the metrics are missin, almost all astarting with couchdb

BR,
Raul

**Comment by gesellix** [comment]
I'll need to check that one locally, thanks for the example view!

**Comment by raulmartinezr** [comment]
Welcome! 

Please, give me a ping when solved so I can test it and go further with version 3.10

Thanks!
Raúl

**Comment by gesellix** [comment]
Minor update: the issue might be caused by the _partitioned_ feature.

**Comment by gesellix** [comment]
This one happens under the hood, I'll also need to improve the error logs to make such errors visible:

```
query_parse_error, reason: `partition` parameter is mandatory for queries to this view.
```

Support for partitioned views will need a bit more time. I'm considering to disable stats for partitioned databases until then. What do you think?

**Comment by raulmartinezr** [comment]
Hi!

I tried and stats are now available in grafana. Here the logs; except first two, the rest seems to be the filter for partitioned databases.


```bash
E0824 23:06:59.959517       1 collector.go:395] Error collecting stats: error collecting couchdb stats: error unmarshalling stats: json: cannot unmarshal array into Go value of type lib.StatsResponse
E0824 23:07:04.953459       1 collector.go:395] Error collecting stats: error collecting couchdb stats: error unmarshalling stats: json: cannot unmarshal array into Go value of type lib.StatsResponse
E0824 23:07:09.967264       1 couchdb-client.go:436] partitioned database /test_given_a_database_when_create_account_then_expected_values currently not supported for view stats
E0824 23:07:09.967293       1 couchdb-client.go:436] partitioned database /test_given_a_database_when_create_account_then_expected_values currently not supported for view stats
E0824 23:07:09.967302       1 couchdb-client.go:436] partitioned database /test_given_a_database_when_create_account_then_expected_values currently not supported for view stats

```

BR,
Raul

**Comment by gesellix** [comment]
Wow. The first two are surprising to me. I'll have to check whether they cause new issues.

**Comment by gesellix** [comment]
@raulmartinezr can you please check again with the updated preview (`docker pull gesellix/couchdb-prometheus-exporter:partitioned-dbs`)?

I added more detailed log messages to find the cause for the error logs you showed in your last comment. Though I couldn't reproduce the errors, I hope the node names might help you to manually execute the related request (`http://couchdb/_node/{node_name}/_stats`, as documented at https://docs.couchdb.org/en/master/api/server/common.html#node-node-name-stats) and then provide the response here.

**Comment by gesellix** [comment]
@raulmartinezr did you have a chance to make another test? I'd like to find the cause for the error messages in your last comment.

**Comment by raulmartinezr** [comment]
Hi!

I'm back. 


I see the same errors
```
E0916 14:43:34.527730       1 collector.go:395] Error collecting stats: error collecting couchdb stats: error unmarshalling stats for node nonode@nohost: json: cannot unmarshal array into Go value of type lib.StatsResponse
E0916 14:43:39.512521       1 collector.go:395] Error collecting stats: error collecting couchdb stats: error unmarshalling stats for node nonode@nohost: json: cannot unmarshal array into Go value of type lib.StatsResponse
```

Maybe it's because my couchdb  is configured as single node?
Stats are reached at http://127.0.0.1:5984/_node/_local/_stats

But it seems stats are collected

![image](https://user-images.githubusercontent.com/4292375/93354536-63d7c500-f83d-11ea-8650-49af0d7ad91e.png)


BR,
Raul

**Comment by gesellix** [comment]
Thanks! I could reproduce it now and it seems like the error messages only appear when the Prometheus exporter queries the CouchDB instance before it's ready. Can you confirm that you scrape the exporter (so, the CouchDB instance, too) while the CouchDB instance is still starting up?

**Comment by gesellix** [comment]
I made some minor tweaks (hide the logging about partitioned views). The release is available as [v29.0.0](https://github.com/gesellix/couchdb-prometheus-exporter/releases/tag/v29.0.0). Thanks a lot for your feedback and participating in the "beta" phase :)
